pkg/R/serialize.R

```diff
@@ -0,0 +1,121 @@
+# Utility functions to serialize R objects so they can be read in Java.
+
+# Type mapping from R to Java
+#  
+# integer -> Int
+# character -> String
+# logical -> Boolean
+# double, numeric -> Double
+# raw -> Array[Byte]
+#
+# list[T] -> Array[T], where T is one of above mentioned types
+# environment -> Map[String, T], where T is a native type
+# jobj -> Object, where jobj is an object created in the backend
+
+writeObject <- function(con, object, withType = TRUE) {
+  # NOTE: In R vectors have same type as objects. So we don't support
+  # passing in vectors as arrays and instead require arrays to be passed
+  # as lists.
+  if (withType) {
+    writeType(con, class(object))
+  }
+  switch(class(object),
+    integer = writeInt(con, object),
+    character = writeString(con, object),
+    logical = writeBoolean(con, object),
+    double = writeDouble(con, object),
+    numeric = writeDouble(con, object),
+    raw = writeRaw(con, object),
+    list = writeList(con, object),
+    jobj = writeString(con, object$id),
+    environment = writeEnv(con, object),
+    stop("Unsupported type for serialization"))
+}
+
+writeString <- function(con, value) {
+  writeInt(con, as.integer(nchar(value) + 1))
+  writeBin(value, con, endian="big")
+}
+
+writeInt <- function(con, value) {
+  writeBin(as.integer(value), con, endian="big")
+}
+
+writeDouble <- function(con, value) {
+  writeBin(value, con, endian="big")
+}
+
+writeBoolean <- function(con, value) {
+  # TRUE becomes 1, FALSE becomes 0
+  writeInt(con, as.integer(value))
+}
+
+writeRawSerialize <- function(outputCon, batch) {
+  outputSer <- serialize(batch, ascii = FALSE, conn = NULL)
+  writeRaw(outputCon, outputSer)
+}
+
+writeRaw <- function(con, batch) {
+  writeInt(con, length(batch))
+  writeBin(batch, con, endian="big")
+}
+
+writeType <- function(con, class) {
+  type <- switch(class,
+            integer = "i",
+            character = "c",
+            logical = "b",
+            double = "d",
+            numeric = "d",
+            raw = "r",
+            list = "l",
+            jobj = "j",
+            environment = "e",
+            stop("Unsupported type for serialization"))
+  writeBin(charToRaw(type), con)
+}
+
+# Used to pass arrays where all the elements are of the same type
+writeList <- function(con, arr) {
+  # All elements should be of same type
+  elemType <- unique(sapply(arr, function(elem) { class(elem) }))
+  stopifnot(length(elemType) <= 1)
+
+  # Write empty lists as strings ?
+  if (length(elemType) == 0) {
+    elemType <- "character"
```

Comment:
It needs to be a valid typename as the Java side does a switch based on the typename. The trouble is that if somebody is passing an empty list we have no idea what the type should be on the java side. Some of this is only working right now as functions take in `Array[Object]` on the java side. 

This seems to be a general problem where if we don't have any elements we can't create the right array type on the Java side. I've written a better TODO for now